Initial SubGhz development

QuantumEF opened the pull request:
Initial work, opening draft pull request to solicit feedback and comments.

This is not yet finished as I have not yet implemented the builder for the CustomSubGhzPreset to allow for custom presets.

QuantumEF commented:
With this latest commit, the following basic app does work and does not crash.

```rust
//! Template project for Flipper Zero.
//! This app prints "Hello, Rust!" to the console then exits.

#![no_main]
#![no_std]

// Required for panic handler
// use flipperzero_rt;

extern crate flipperzero_alloc;

use core::ffi::CStr;

use flipperzero::dialogs::alert;
use flipperzero::println;
use flipperzero::subghz;
use flipperzero_rt::{entry, manifest};

// Define the FAP Manifest for this application
manifest!(
    name = "Flipper Zero Rust",
    app_version = 1,
    has_icon = true,
    // See https://github.com/flipperzero-rs/flipperzero/blob/v0.11.0/docs/icons.md for icon format
    icon = "rustacean-10x10.icon",
);

// Define the entry function
entry!(main);

// Entry point
fn main(_args: Option<&CStr>) -> i32 {
    alert("Hello Erik");

    let dev_name = c"cc1101_int";

    let x = subghz::SubGhz::subghz_devices_get_by_name(dev_name);

    match x {
        Some(_) => alert("YEA"),
        None => alert("Boooo"),
    }

    0
}
```

QuantumEF commented:
@dcoles Do you have time to give feedback on what I have done so far? ie: just how I have started going about doing things.
As mentioned in my previous comment, I have only tested just opening device. I am continuing testing to validate the functionality of other things.

JarvisCraft commented on crates/flipperzero/Cargo.toml:
```diff
@@ -41,6 +41,7 @@ embedded-hal-0 = { package = "embedded-hal", version = "0.2.7", features = ["unp
 
 # Docs
 document-features = { workspace = true, optional = true }
+uom = { version = "0.36.0", features = ["si", "u32"], default-features = false }
```
I suggest that we add all new dependencies to workspace root and reference them here via `workspace = true`

Also note, that this part of `Cargo.toml` is for docs-related dependencies (see comment two lines above)

QuantumEF replied:
Makes sense to me, I will hold off on this change for now since I need to properly figure out how to use uom crate since at present my use of it utilizes dynamic dispatch. I also want to read up on cargo workspaces since I have not used them before.

JarvisCraft commented on crates/flipperzero/src/subghz/device.rs:
```diff
@@ -0,0 +1,193 @@
+
+use super::{error::SubGhzError, preset::SubGhzPreset};
+use core::{ffi::CStr, ptr::null_mut};
+use flipperzero_sys::{
```
I feel like importing every function from `sys` only to call it once complicates this source so I suggest aliasing `flipperzero_sys` as `sys` and using the qualified paths, i.e.:

```rust
use flipperzero_sys as sys;
```

and

```rust
let result = unsafe { sys::subghz_devices_begin(self.device) };
```

JarvisCraft commented on crates/flipperzero/src/subghz/device.rs:
```diff
@@ -0,0 +1,193 @@
+}
+
+impl SubGhz {
+    /// For the internal device use "cc1101_int" (From `lib/subghz/devices/cc1101_int/cc1101_int_interconnect.c`)
```
A link would be useful, I think

Note that there is cool Markdown syntax for link in footers:

```suggestion
    /// For the internal device use [`cc1101_int`]
    ///
    /// [`cc1101_int`]: https://github.com/flipperdevices/flipperzero-firmware/blob/dev/lib/subghz/devices/cc1101_int/cc1101_int_interconnect.c
```

JarvisCraft commented on crates/flipperzero/src/subghz/device.rs:
```diff
@@ -0,0 +1,193 @@
+    pub fn subghz_devices_get_by_name(name: &CStr) -> Option<SubGhz> {
+        unsafe { subghz_devices_init() }
+
+        // Safety: input type enforeced by CStr type.
```
I am not sure that this is sound: `CStr` only guarantees pointer validity and the absence of `\0`s, but it does *not* guarantee that the string is valid UTF-8 and (but I may be mistaken here) some Furi APIs require more specific strings.

QuantumEF replied:
Based on what I saw, the string is eventually used by the following function: https://github.com/flipperdevices/flipperzero-firmware/blob/release/lib/subghz/registry.c#L3

My interpretation is that I just needed to make sure I had a null terminated array as the input. That being said, I was looking at commit 10 behind that I linked to, so I should probably look again since it looks like the file changed a fair bit.

QuantumEF replied:
To add additional info, the string is compared with strcmp() to find the matching device in the list.

And no, CStr does not guarantee valid UTF-8 (it can be created from any &[u8] and the "constructors" only are checking for null termination.)

JarvisCraft replied:
Nice, given that it is only used for `strcmp`, `CStr` fits the purpose

JarvisCraft commented on crates/flipperzero/src/subghz/device.rs:
```diff
@@ -0,0 +1,193 @@
+        // Safety: This pointer should either be null or return a reference, I don't see how it could not.
+        // Further, the reference should be static as it relates to the variable:
+        // `static SubGhzDeviceRegistry* subghz_device_registry = NULL;` in `lib/subghz/devices/registry.c``
+        let dev = unsafe { dev.as_ref()? };
```
Let's move `?` out of `unsafe`

JarvisCraft commented on crates/flipperzero/src/subghz/device.rs:
```diff
@@ -0,0 +1,193 @@
+        // Further, the reference should be static as it relates to the variable:
+        // `static SubGhzDeviceRegistry* subghz_device_registry = NULL;` in `lib/subghz/devices/registry.c``
+        let dev = unsafe { dev.as_ref()? };
+        Some(Self { device: dev })
```
Probably worth renaming the variable to `device` to use shorthand form here:

```suggestion
        Some(Self { device })
```

JarvisCraft commented on crates/flipperzero/src/subghz/device.rs:
```diff
@@ -0,0 +1,193 @@
+    // No clue what sort of error may occur, if using the internal device, it will always return Ok()
+    pub fn begin(&mut self) -> Result<(), SubGhzError> {
+        // Safety: self.device is not Null so this will not crash when furi_check is invoked.
+        let result = unsafe { subghz_devices_begin(self.device) };
```
`failed` seems to be a more easy-to-review name for the variable

JarvisCraft commented on crates/flipperzero/src/subghz/device.rs:
```diff
@@ -0,0 +1,193 @@
+
+    // No clue what sort of error may occur, if using the internal device, it will always return Ok()
+    pub fn begin(&mut self) -> Result<(), SubGhzError> {
+        // Safety: self.device is not Null so this will not crash when furi_check is invoked.
```
Important: `furi_check` causing a crash is *not* unsafe thus it should not be mentioned in safety comments (in the same matter as `panic!` is safe to call), although mentioning non-nullity is useful.

What should be mentioned here is if there are any `*const SubGhzDevice` invariants which should be upheld.

Same goes for the following `unsafe` calls in the file.

QuantumEF replied:
I am aware that a crash is considered safe; as you mentioned I wanted to put it there to mention the non-nullity.

This is something I want to look into more, so thanks for your feedback on this. I will mark this comment as resolved as you have more details in your other comment.

QuantumEF replied:
As a sidenote, I think I lost a bunch of replys to some of your comments since I clicked "Resolve Conversation" without adding the comment :/ I will go back and try add those comments, but most of them were just links to the commits in which I fixed them.

JarvisCraft commented on crates/flipperzero/src/subghz/device.rs:
```diff
@@ -0,0 +1,193 @@
+use uom::si::{frequency::hertz, u32::Frequency};
+
+pub struct SubGhz {
+    device: &'static SubGhzDevice,
```
Storing `*const SubGhzDevice` as a Rust reference is really tricky, considering that there are much more invariant to uphold for it, especially given that `SubGhzDevice` seems to have some form of interior mutability internally. I suggest that you store `NonNull<SubGhzDevice>` instead, only creating short-living references when needed (e.g. when reading fields).

QuantumEF replied:
So looking at [this instance of a SubGhzDevice](https://github.com/flipperdevices/flipperzero-firmware/blob/fe424061c968c41131448697ea587533402dccae/lib/subghz/devices/cc1101_int/cc1101_int_interconnect.c#L93), it looks like a given SubGhzDevice is going to be a struct that contains a name and a pointer to a struct that contains a bunch of function pointers, so I don't think this is an issue; but I also do not have much practical experience writing in C, so I could not be interpreting things correctly. All that being said, making the change back to `*const SubGhzDevice` is simple and only effect the `get_rssi` ang `get_lqi` functions, though I am not sure how should now change the semantics of those functions: ie should they still be `Option<T>` or `Result<T>` to indicate if the device pointer is null or if the function pointer is null.

JarvisCraft commented on crates/flipperzero/src/subghz/device.rs:
```diff
@@ -0,0 +1,193 @@
+    }
+
+    pub fn read_packet(&mut self, buf: &mut [u8]) -> usize {
+        let mut size = if buf.len() > 255 {
+            255_u8
+        } else {
+            buf.len() as u8
+        };
```
```suggestion
        let mut size = u8::try_from(buf.len()).unwrap_or(u8::MAX);
```

JarvisCraft commented on crates/flipperzero/src/subghz/device.rs:
```diff
@@ -0,0 +1,193 @@
+        // Safety: self.device is not Null so this will not crash when furi_check is invoked.
+        unsafe {
+            subghz_devices_end(self.device);
+            subghz_devices_deinit();
```
What happens if multiple instances of `SubGhz` are created and thus `subghz_devices_deinit` is called (on ones destructor) while the other one is still active?

I guess we may need to address this upstream, e.g. request adding a counter to it (e.g. the number of currently active *activations*)

QuantumEF replied:
In some haphazard test code:

```rust
fn main(_args: Option<&CStr>) -> i32 {
    let dev_name = c"cc1101_int";

    let Some(mut x) = subghz::SubGhz::subghz_devices_get_by_name(dev_name) else {
        alert("BOOOO");
        return 0;
    };

    x.load_preset(subghz::preset::SubGhzPreset::Ook270Async);

    let Ok(()) = x.begin() else {
        return 0;
    };

    let Ok(actual_freq) = x.set_frequency(Frequency::new::<hertz>(433_500_000)) else {
        return 0;
    };

    let mut s = String::<128>::new();

    alert("about to take second ref");
    // Crash
    let Some(y) = subghz::SubGhz::subghz_devices_get_by_name(dev_name) else {
        alert("Cant take second reference");
        return 0;
    };

    0
}
```

Attempting to create a second reference crashes the device (I am not too surprised since the API does a lot of checking for safety). That being said, I don't know which function actually crashes things, so I don't know how it is enforced: I would like to be able to enforce this to prevent the crash.

JarvisCraft commented on crates/flipperzero/src/subghz/device.rs:
```diff
@@ -0,0 +1,193 @@
+
+    pub fn write_packet(&mut self, buf: &[u8]) -> Result<(), SubGhzError> {
+        if buf.len() > 255 {
+            return Err(SubGhzError::PacketTooLong);
```
I suggest adding `len` field to this error variant

QuantumEF replied:
Good idea. I should also probably do this with the other variants as well. For now I will just do it for the PacketTooLong variant

JarvisCraft commented on crates/flipperzero/src/subghz/preset.rs:
```diff
@@ -0,0 +1,64 @@
+pub struct CustomSubGhzPreset<'a>(&'a [u8]);
+
+impl<'a> CustomSubGhzPreset<'a> {
+    /// # Safety
```
```suggestion
    /// # Safety
    ///
```


Markdown requires a newline after a header

JarvisCraft commented on crates/flipperzero/src/subghz/preset.rs:
```diff
@@ -0,0 +1,64 @@
+    }
+}
+
+#[derive(Clone, Debug)]
+/// This struct is defined for creating custom configurations. It enforces the null padding needed at the end. along with the power amplifier config?
```
```suggestion
/// This struct is defined for creating custom configurations. It enforces the null padding needed at the end. along with the power amplifier config?
#[derive(Clone, Debug)]
```

QuantumEF replied:
Changed with a3036929a2a702be8313d927819880dd94544a85
(Was not able to use the "Commit Suggestion" feature since the change is "outdated" due to all the other changes I have committed as I have been going through your suggestions.)

JarvisCraft commented on crates/flipperzero/src/subghz/preset.rs:
```diff
@@ -0,0 +1,64 @@
+
+impl<'a> CustomSubGhzPreset<'a> {
+    /// # Safety
+    /// Read the manual for the CC1101 chip and review the [`furi_hal_subghz_load_custom_preset()`](https://github.com/flipperdevices/flipperzero-firmware/blob/0eaad8bf64f01a6f932647a9cda5475dd9ea1524/targets/f7/furi_hal/furi_hal_subghz.c#L162) function provided by the flipper API
```
```suggestion
    /// Read the manual for the CC1101 chip and review the [`furi_hal_subghz_load_custom_preset()`].
    ///
    /// [`furi_hal_subghz_load_custom_preset()`]: (https://github.com/flipperdevices/flipperzero-firmware/blob/0eaad8bf64f01a6f932647a9cda5475dd9ea1524/targets/f7/furi_hal/furi_hal_subghz.c#L162) function provided by the flipper API
```

QuantumEF replied:
Not quite sure what exactly the change is here. My assumption is for it to include both a link to the function in the `flipperzero_sys::furi_hal_etc` documentation as well as the link I provide to the C source code?

JarvisCraft replied:
My bad, I've accidentally trimmed the end of the comment sentence.

The intention here is to move the URL to the markdown footer so that the header text is easier to read

JarvisCraft commented on crates/flipperzero/src/subghz/preset.rs:
```diff
@@ -0,0 +1,64 @@
+}
+
+impl<'a> SubGhzPreset<'a> {
+    pub fn into_furi_preset(&self) -> FuriHalSubGhzPreset {
```
Not sure, but maybe its worth adding `impl From<&SubGhzPreset<'_>> for FuriHalSubGhzPreset>`

Also, `as_` prefix [seems to be][1] more valid for the function

[1]: https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv

QuantumEF replied:
I am probably going to hold off for now on implementing this since [the docs](https://doc.rust-lang.org/std/convert/trait.From.html) mention that "The From trait is intended for perfect conversions," This function looses data when converting from the `Custom` variant.

I like the `as_` suggestion because it is something I always forget the semantics about (I should print it out, lol).
When I originally created some of the structs, I was not able to implement copy, but now I am  since the custom variant is `Custom(&[u8])`. Because of this, I am wondering if I should just change the function signature to take in `self` instead of a `&self` then maybe implying is should use `to_` ?

JarvisCraft commented on crates/flipperzero/src/subghz/preset.rs:
```diff
@@ -0,0 +1,64 @@
+impl<'a> SubGhzPreset<'a> {
+    pub fn into_furi_preset(&self) -> FuriHalSubGhzPreset {
+        match self {
+            SubGhzPreset::IDLE => FuriHalSubGhzPreset_FuriHalSubGhzPresetIDLE,
+            SubGhzPreset::Ook270Async => FuriHalSubGhzPreset_FuriHalSubGhzPresetOok270Async,
+            SubGhzPreset::Ook650Async => FuriHalSubGhzPreset_FuriHalSubGhzPresetOok650Async,
+            SubGhzPreset::_2FSKDev238Async => {
+                FuriHalSubGhzPreset_FuriHalSubGhzPreset2FSKDev238Async
+            }
+            SubGhzPreset::_2FSKDev476Async => {
+                FuriHalSubGhzPreset_FuriHalSubGhzPreset2FSKDev476Async
+            }
+            SubGhzPreset::MSK99_97KbAsync => FuriHalSubGhzPreset_FuriHalSubGhzPresetMSK99_97KbAsync,
+            SubGhzPreset::GFSK9_99KbAsync => FuriHalSubGhzPreset_FuriHalSubGhzPresetGFSK9_99KbAsync,
+            SubGhzPreset::Custom(_) => FuriHalSubGhzPreset_FuriHalSubGhzPresetCustom,
```
```suggestion
            Self::IDLE => FuriHalSubGhzPreset_FuriHalSubGhzPresetIDLE,
            Self::Ook270Async => FuriHalSubGhzPreset_FuriHalSubGhzPresetOok270Async,
            Self::Ook650Async => FuriHalSubGhzPreset_FuriHalSubGhzPresetOok650Async,
            Self::_2FSKDev238Async => {
                FuriHalSubGhzPreset_FuriHalSubGhzPreset2FSKDev238Async
            }
            Self::_2FSKDev476Async => {
                FuriHalSubGhzPreset_FuriHalSubGhzPreset2FSKDev476Async
            }
            Self::MSK99_97KbAsync => FuriHalSubGhzPreset_FuriHalSubGhzPresetMSK99_97KbAsync,
            Self::GFSK9_99KbAsync => FuriHalSubGhzPreset_FuriHalSubGhzPresetGFSK9_99KbAsync,
            Self::Custom(_) => FuriHalSubGhzPreset_FuriHalSubGhzPresetCustom,
```

JarvisCraft commented on crates/flipperzero/src/subghz/preset.rs:
```diff
@@ -0,0 +1,64 @@
+use flipperzero_sys::{
```
Same as preciously, a single `use flipperzero_sys as sys` would look cleaner

QuantumEF replied:
Changed with 37c523b869a3a454792d2ff8c6b23d5d0c6f6f31

JarvisCraft commented on crates/flipperzero/src/subghz/error.rs:
```diff
@@ -0,0 +1,22 @@
+use core::{error::Error, fmt::Display};
+
+#[derive(Clone, Copy, Debug)]
```
`uDebug` should also be implemented when `Debug` is implemented.

QuantumEF replied:
I didn't realize this, added with 1ebd7ca9132f77ced2fcc554628f39cdac56c5d6

JarvisCraft reviewed (changes requested):
The first iteration of the changes

They maybe nitpicky, but I wish that they will be helpful

QuantumEF commented:
> The first iteration of the changes
> 
> They maybe nitpicky, but I wish that they will be helpful

Thank you so much for your feedback and help! As you see I have gone through most of your suggestions and added them. The ones I have not marked as resolved yet are because I did want to think on them and possible discuss them more. 

In my error, I lost some of the comments on the changes I marked as resolved because I assumed the "Marked as resolved" button did something similar to the "Comment and mark as resolved" button I have seen elsewhere; Oops, most of those comments were just referencing the commit I changed things in.

JarvisCraft commented:
> Thank you so much for your feedback and help!

You are welcome! Thank you for working on this feature!

> As you see I have gone through most of your suggestions and added them. The ones I have not marked as resolved yet are because I did want to think on them and possible discuss them more.

For sure! I understand that this is a draft, so my review should be considered as more of the *recommendations* by now :)

> In my error, I lost some of the comments on the changes I marked as resolved because I assumed the "Marked as resolved" button did something similar to the "Comment and mark as resolved" button I have seen elsewhere; Oops, most of those comments were just referencing the commit I changed things in.

No problems, GitHub is a bit weird with PR review threads anyway

JarvisCraft commented:
It may take me some time to continue the review (I will be at the conference for a few days) thus feel free to leave some questions unresolved as-is since eventually the will be reviewed
